api/api.go

```diff
@@ -0,0 +1,387 @@
+	resp interface{}) error {
+
+	return c.DoWithHeaders(
+		ctx, http.MethodDelete, path, headers, nil, resp)
+}
+
+func (c *client) Do(
+	ctx context.Context,
+	method, path string,
+	body, resp interface{}) error {
+
+	return c.DoWithHeaders(ctx, method, path, nil, body, resp)
+}
+
+func beginsWithSlash(s string) bool {
+	return s[0] == '/'
+}
+
+func endsWithSlash(s string) bool {
+	return s[len(s)-1] == '/'
+}
+
+func (c *client) DoWithHeaders(
+	ctx context.Context,
+	method, uri string,
+	headers map[string]string,
+	body, resp interface{}) error {
+
+	res, err := c.DoAndGetResponseBody(
+		ctx, method, uri, headers, body)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// parse the response
+	switch {
+	case res == nil:
+		return nil
+	case res.StatusCode >= 200 && res.StatusCode <= 299:
+		if resp == nil {
+			return nil
+		}
+		dec := json.NewDecoder(res.Body)
+		if err = dec.Decode(resp); err != nil && err != io.EOF {
+			log.WithError(err).Errorf(
+				"Unable to decode response into %+v", resp)
+			return err
+		}
+	default:
+		return c.ParseJSONError(res)
+	}
+
+	return nil
+}
+
+func (c *client) DoAndGetResponseBody(
+	ctx context.Context,
+	method, uri string,
+	headers map[string]string,
+	body interface{}) (*http.Response, error) {
+
+	var (
+		err                error
+		req                *http.Request
+		res                *http.Response
+		ubf                = &bytes.Buffer{}
+		luri               = len(uri)
+		hostEndsWithSlash  = endsWithSlash(c.host)
+		uriBeginsWithSlash = beginsWithSlash(uri)
+	)
+
+	ubf.WriteString(c.host)
+
+	if !hostEndsWithSlash && (luri > 0) {
+		ubf.WriteString("/")
+	}
+
+	if luri > 0 {
+		if uriBeginsWithSlash {
+			ubf.WriteString(uri[1:])
+		} else {
+			ubf.WriteString(uri)
+		}
+	}
+
+	u, err := url.Parse(ubf.String())
+	if err != nil {
+		return nil, err
+	}
+
+	var isContentTypeSet bool
+
+	// marshal the message body (assumes json format)
+	if body != nil {
+		if r, ok := body.(io.ReadCloser); ok {
+			req, err = http.NewRequest(method, u.String(), r)
+			defer r.Close()
+			if v, ok := headers[HeaderKeyContentType]; ok {
+				req.Header.Set(HeaderKeyContentType, v)
+			} else {
+				req.Header.Set(
+					HeaderKeyContentType, headerValContentTypeBinaryOctetStream)
+			}
+			isContentTypeSet = true
+		} else {
+			buf := &bytes.Buffer{}
+			enc := json.NewEncoder(buf)
+			if err = enc.Encode(body); err != nil {
+				return nil, err
+			}
+			req, err = http.NewRequest(method, u.String(), buf)
+			if v, ok := headers[HeaderKeyContentType]; ok {
+				req.Header.Set(HeaderKeyContentType, v)
+			} else {
+				req.Header.Set(HeaderKeyContentType, HeaderValContentTypeJSON)
+			}
+			isContentTypeSet = true
+		}
+	} else {
+		req, err = http.NewRequest(method, u.String(), nil)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !isContentTypeSet {
+		isContentTypeSet = req.Header.Get(HeaderKeyContentType) != ""
+	}
+
+	// add headers to the request
+	if len(headers) > 0 {
```

Comment:
Hi @codenrhoden,

The outer level check and indentation are unnecessary. Consider rewriting it without `if len(headers) > 0 {`:

```go
	for header, value := range headers {
 		if header == HeaderKeyContentType && isContentTypeSet {
 			continue
 		}
 		req.Header.Add(header, value)
 	}
```